Add options to `fast-filedeps` to bring it close to parity with V1 `filedeps`

Eric-Arellano opened the pull request:
### Problem
We want to finish replacing V1 `filedeps` with V2 `fast-filedeps`. Keeping both goals around gives us added complexity, e.g. more time in CI for tests.

To finish the replacement, `fast-filedeps` must first get support for `--absolute` and `--globs`. 

### Solution
Add support for `--absolute` and `--globs`.

Also, fix and refactor the unit tests, which were skipped in https://github.com/pantsbuild/pants/pull/6880 and never restored. Update the integration test to mirror the V1 test.

We can't yet actually drop the V1 goal yet due to two features that need to be implemented for full feature parity: 1) handling `scala_library`'s `java_sources` and 2) supporting `jvm_app`s. This PR adds failing unit tests for both those cases. Once those tests can be un-skipped, then we can drop V1.

stuhood commented on src/python/pants/rules/core/filedeps.py:
```diff
@@ -16,24 +17,36 @@ class Filedeps(LineOriented, Goal):
+  absolute = filedeps_options.values.absolute
+
+  unique_rel_paths = set()
+  build_root = get_buildroot()
```
So... I think that this is fine. But we should consider whether we want to DI the buildroot the same way we DI most everything in v2. The easiest thing would probably be to make it a singleton at engine-creation time: https://github.com/pantsbuild/pants/blob/da0989f9f00cf2a9eca58410a721d23a1fdf72bb/src/python/pants/init/engine_initializer.py#L329-L339

As you've seen, having this be a global mutable variable can be painful from a testing perspective, and we should look for opportunities to minimize that pain in v2.

Eric-Arellano replied:
Hm, this sounds good to me but I don't really know how to approach this. Should we add a TODO or would someone have a few minutes to add a commit with this? I can try with a bit more direction also.

stuhood replied:
AFAICT, the sum total of code involved would be:
```
 @rule(BuildRoot, []) 
 def build_root_singleton(): 
   return BuildRoot.instance
  # ... a few lines down in that file
  rules += [build_root_singleton]

# in a rule:
@rule(.., [.., BuildRoot])
```

This is not a big deal either way though.

illicitonion replied:
I was kind of hoping we'd just deprecate the absolute flag, rather than need to worry about this. WDYT?

Eric-Arellano replied:
Implemented the rule Stu proposed. Don't want to wait for the deprecation cycle and I think this rule could be helpful in the future.

Thanks for the suggestion!

stuhood commented on src/python/pants/rules/core/filedeps.py:
```diff
@@ -16,24 +17,36 @@ class Filedeps(LineOriented, Goal):
   """
 
-  # TODO: Until this implements more of the options of `filedeps`, it can't claim the name!
+  # TODO: Until this implements `--globs`, this can't claim the name `filedeps`!
```
I think that to implement `--globs`, this could inspect `sources.filespec`, potentially?

Eric-Arellano replied:
That worked! Thanks!

cattibrie commented on tests/python/pants_test/rules/test_filedeps_integration.py:
```diff
@@ -1,35 +1,44 @@
-      examples/src/resources/org/pantsbuild/example/hello/world.txt
-      ''')
-    )
+  def test_filedeps_globs(self):
```
It would be nice to update `examples` to have at least two files of the same extension (like .scala) in one directory to test that `--globe` option works for more than one file.

cattibrie commented:
@Eric-Arellano Thanks for this PR! Really like type annotations and how tests look now!

stuhood commented on src/python/pants/rules/core/filedeps.py:
```diff
@@ -16,28 +17,47 @@ class Filedeps(LineOriented, Goal):
 def file_deps(
   console: Console,
   filedeps_options: Filedeps.Options,
+  build_root: BuildRoot,
```
So... one development since this was opened is the `Workspace` type.

I _think_ that the semantics of the `BuildRoot` type (exposes some information) are sufficiently different from the `Workspace` type (implies mutable access) that having them separate could be reasonable. I could also see merging them, and exposing `Workspace.root` or something. "Relatively small single purpose things to DI" is a thing... but the `BuildRoot` is almost too specific a thing, and would always have a subset of the fields of `Workspace`.

Fine either way though.

stuhood replied:
cc @gshuflin

Eric-Arellano commented:
@stuhood could you please re-review? We can't drop V1 yet because we don't properly handle two JVM related use cases :/ I updated the PR description and added skipped failing unit tests.

This does at least bring us much closer to dropping the V1 goal!
